### PR TITLE
[KT-57607] Fix performance of ArrayList.addAll methods in JS stdlib

### PR DESCRIPTION
**Pull request by DeoTimeTheGithubUser**
Relating issue: [KT-57607](https://youtrack.jetbrains.com/issue/KT-57607/MutableList.addAll-has-UNACCEPTABLE-performance-on-Kotlin-JS-legacy-IR)

### CODE REVIEW

File: libraries/stdlib/js/src/kotlin/collections/ArrayList.kt
Comment (ilya-g) [inline libraries/stdlib/js/src/kotlin/collections/ArrayList.kt:74]: AFAIK, there was some limitation on how many parameters can be spread using this approach, so it may fail on large `elements` collection

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/js/src/kotlin/collections/ArrayList.kt
Comment (robledo6) [reply libraries/stdlib/js/src/kotlin/collections/ArrayList.kt:167]: Nasa. Más

**Comment by DeoTimeTheGithubUser** [comment]
Thanks, I've changed it to iterating over the elements instead of spreading into a method

File: libraries/stdlib/js/src/kotlin/collections/ArrayList.kt
Comment (ilya-g) [inline libraries/stdlib/js/src/kotlin/collections/ArrayList.kt:75]: Would be better to expand the array by setting its length and then fill elements manually.

File: libraries/stdlib/js/src/kotlin/collections/ArrayList.kt
Comment (ilya-g) [inline libraries/stdlib/js/src/kotlin/collections/ArrayList.kt:89]: Moving array tail on each element insertion would result in quadratic complexity.

File: libraries/stdlib/js/src/kotlin/collections/ArrayList.kt
Comment (ilya-g) [inline libraries/stdlib/js/src/kotlin/collections/ArrayList.kt:154]: It doesn't seem that jsArrayPrototype is now used anywhere except this function. Does it make sense to extract it then to a variable?

**Review by ilya-g — CHANGES_REQUESTED** [review]

**Comment by DeoTimeTheGithubUser** [comment]
Sorry for the delay, will get to these tmr

**Review by ilya-g — APPROVED** [review]
Neat. Let's wait for a CI run and I'll merge it after that.

**Comment by LouisCAD** [comment]
Can we have a benchmark to compare the performance before/after, and compared to calling add multiple times?

**Comment by ilya-g** [comment]
@LouisCAD I'd leave that as an exercise for the reader :)

**Comment by LouisCAD** [comment]
Is there a directory for existing Kotlin/JS benchmarks where I could submit a PR?

I'd also be curious about if skipping splice would have any beneficial effect performance wise when the index is zero, and how much the impact would be for the non zero case.

**Comment by ilya-g** [comment]
No, we don't have a directory with stdlib benchmarks.
Usually we create a one-off repository when we need to benchmark some alternatives where we compare different implementations. For benchmarking Kotlin/JS, we use `kotlinx.benchmark`. For example: https://github.com/qurbonzoda/StringBuilderBenchmarks/blob/master/build.gradle

**Pull request merged**
